File: src/spirite/base/graphics/renderer/sources/RenderSource.java
package spirite.base.graphics.renderer.sources;

import spirite.base.graphics.RawImage;
import spirite.base.graphics.renderer.RenderEngine.RenderSettings;
import spirite.base.image_data.GroupTree.Node;
import spirite.base.image_data.ImageWorkspace;
import spirite.base.image_data.MediumHandle;

import java.util.ArrayList;
import java.util.List;

/**
 * A RenderSource corresponds to an object which can be rendered and it implements
 * everything needed to perform a Render using certain RenderSettings.
 *
 * Note: It is important that subclasses overload the equals and hashCode methods
 * of each RenderSource since the RenderEngine uses them to determine if you
 * are rendering the same thing as something that has already been rendered.
 * If you just go on the built-in uniqueness test and pass them through renderImage
 * then unless you are storing the RenderSource locally yourself (which is possible
 * and not harmful but defeats the purpose of RenderEngine), then RenderEngine 
 * will get clogged remembering different renders of the same image.
 */
public abstract class RenderSource {
	public final ImageWorkspace workspace;
	RenderSource( ImageWorkspace workspace) {this.workspace = workspace;}
	public abstract int getDefaultWidth();
	public abstract int getDefaultHeight();
	public abstract List<MediumHandle> getImagesReliedOn();
	public List<Node> getNodesReliedOn() {return new ArrayList<>(0);}
	public abstract RawImage render( RenderSettings settings);
	
	@Override
	public int hashCode() {
		final int prime = 31;
		int result = 1;
		result = prime * result + ((workspace == null) ? 0 : workspace.hashCode());
		return result;
	}
	@Override
	public boolean equals(Object obj) {
		if (this == obj)
			return true;
		if (obj == null)
			return false;
		if (getClass() != obj.getClass())
			return false;
		RenderSource other = (RenderSource) obj;
		if (workspace == null) {
			if (other.workspace != null)
				return false;
		} else if (!workspace.equals(other.workspace))
			return false;
		return true;
	}
}

File: src/main/kotlin/spirite/base/graphics/rendering/sources/RenderSource.kt
package spirite.base.graphics.rendering.sources

import spirite.base.graphics.GraphicsContext
import spirite.base.graphics.rendering.RenderSettings
import spirite.base.imageData.IImageWorkspace
import spirite.base.imageData.MediumHandle
import spirite.base.imageData.groupTree.GroupTree.*


/**
 * A RenderSource corresponds to an object which can be rendered and it implements
 * everything needed to perform a Render using certain RenderSettings.
 *
 * Note: It is important that subclasses overload the equals and hashCode methods
 * (or better yet be data classes), so that Identical Calls are correctly identified
 * as being identical.
 */
interface RenderSource {
    val workspace: IImageWorkspace
    val defaultWidth: Int
    val defaultHeight: Int
    val imageDependencies: Collection<MediumHandle>
    val nodeDependencies: Collection<Node>

    fun render( settings: RenderSettings, gc: GraphicsContext)
}

fun getRenderSourceForNode( node: Node, workspace: IImageWorkspace) : RenderSource {
    when( node) {
        is GroupNode -> return GroupNodeSource(node, workspace)
        is LayerNode -> return LayerSource(node.layer, workspace)
        else -> throw Exception("Unrecognizd Node Type")
    }
}